### PR TITLE
Update docker runtime build setup

### PR DESCRIPTION
**Pull request by matthias-wende-sociomantic**
* default config file directory is now `/etc/dmqnode/` such that external volumes can be mounted (e.g. by kubernetes)
* provide default config and credentials file in `/etc/dmqnode`
* do not use entrypoint script, thus node runs as PID1 and is able to receive signals from the docker daemon
* move Docker file in the root dir to make the building more convenient
* do not use entrypoint script as it became obsolete. Also the now node runs as PID1 and is able to receive signals from docker daemon
* remove unnecessary files

### CODE REVIEW

**Comment by matthias-wende-sociomantic** [comment]
These are for now the suggestions for the updated docker build files.
The basic changes are that `dmqnode` binds to all network devices, runs as PID1 and that config files are in `/etc/dmqnode`. These are done with having a future kubernetes deployment in mind.

The rest changes are cosmetic.

Last think todo is to update\add the docker-compose example from the dhtnode repo.

File: docker/entrypoint.sh [hunk 1]
```diff
@@ -1,11 +0,0 @@
...
-set -e
-
-if [ -z "$CREDENTIALS" ]; then
-    echo "Must defined CREDENTIALS env var in form of client_name:key"
-    exit 1
-fi;
-
-echo $CREDENTIALS > etc/credentials
-
-dmqnode -c etc/config.ini
```
Comment (Geod24) [inline docker/entrypoint.sh]: You could just use `exec dmqnode` on this line, to keep the script but have dmqnode receive signals.

Comment (matthias-wende-sociomantic) [reply docker/entrypoint.sh]: Hmm, I should have mentioned that the actual reason for removing the script is, that there is just no need for it any longer :). For "real world" uses the etc directory, containing a credentials file, is mounted into the container and this image as it comes provides test credentials.

**Comment by matthias-wende-sociomantic** [comment]
The "putting all shell scripts into Dockerfile itself" is a debatable style consideration. The interesting thing here is that the readability argument could be seen exactly the other way around :), i.e. it's easier to have everything in one file instead of opening other files to see what's happening.

My concern about someone running the scripts on the host by accident, comes from my experience that  scripts that helps building the container (i.e. run `docker build` + parameters) can be mistakenly interchanged with the ones that are run inside the Dockerfile. It's a problem since sometimes these scripts run `sudo apt install`. (I know this isn't the case here).

After all having the scripts inside the `Dockerfile` means less instruction and in fact that means less images. Given the fact that all the other pro/cons are mainly style issues I'd go with in `Dockerfile` scripts.

File: doc/etc/config.ini [hunk 2]
```diff
@@ -25,9 +25,9 @@
 # unix_socket_path      The path of the UNIX domain server socket.
 
 [Server]
-address                 = # address
-port                    = # port for the incoming connections on legacy protocol
-neoport                 = # port for the incoming connections on the neo protocol
+address                 = 0.0.0.0
+port                    = 10000
+neoport                 = 10001
```
Comment (leandro-lucarella-sociomantic) [inline doc/etc/config.ini]: Maybe keep the comments as pure comments above each value? (the options are pretty self explanatory though, so is fine if you want to remove them).

File: docker/Dockerfile.dmqnode [hunk 3]
```diff
@@ -1,26 +1,62 @@
...
+        libebtree6-dev \
+        liblzo2-dev \
+        libglib2.0-dev \
+        libpcre3-dev \
+        libgcrypt-dev \
+        libgpg-error-dev \
+    \
+    # Build app \
+    && if_d 2 make d2conv \
+    && make all pkg F=production
```
Comment (leandro-lucarella-sociomantic) [inline docker/Dockerfile.dmqnode]: Please put this in a separate script, this is awful to maintain and very error prone. At the level of the Dockerfile the only interesting information is that you are building the project[1], all the small details you generally don't care about, the same as you do `make all` instead of writing every single command in the Dockerfile.

[1] Maybe the dependencies too, but you fix that with just leaving the `apt install` in the dockerfile for everything that is not the compiler. If you also want to keep the `F=production` explicit, just do a `RUN F=production ./build`.

File: docker/Dockerfile.dmqnode [hunk 4]
```diff
@@ -1,26 +1,62 @@
...
 WORKDIR /project/
 # Redefine arguments as env vars to be used inside build.sh script:
 ARG DIST=xenial
 ARG DMD=2.078.*
-ENV DMD=$DMD DIST=$DIST
-RUN docker/build.sh
+ENV DMD=$DMD
+RUN set -eu \
+    # Compiler configuration \
+    && . submodules/beaver/lib/dlang.sh \
```
Comment (leandro-lucarella-sociomantic) [inline docker/Dockerfile.dmqnode]: This is referring to beaver which I guess is not a submodule, otherwise you wouldn't have copy&pasted all this code?

File: docker/Dockerfile.dmqnode [hunk 5]
```diff
@@ -1,26 +1,62 @@
...
-COPY docker/install.sh /tmp/
-RUN /tmp/install.sh && rm /tmp/install.sh
-# Update node folder with relevant config to finalize installation
-COPY ./doc/etc/config.docker.ini /srv/dmqnode/dmqnode-0/etc/config.ini
+RUN set -eu \
+    && apt update && apt install -y libebtree6 \
+    && mkdir /etc/dmqnode \
+    && mkdir -p /srv/dmqnode/dmqnode-0/log \
+    && ln -s /etc/dmqnode /srv/dmqnode/dmqnode-0/etc \
+    && apt install -y /packages/*.deb
```
Comment (leandro-lucarella-sociomantic) [inline docker/Dockerfile.dmqnode]: Same here. The only interesting part is you are installing the built package (and the run-time dependencies you want to install).

**Comment by leandro-lucarella-sociomantic** [comment]
BTW, besides those objections, the rest looks good. :)

File: doc/etc/config.ini [hunk 6]
```diff
@@ -69,8 +69,8 @@ write_index_ms = 60000
 #                       enables outputting DMQ channel throughput statistics to
 #                       the console.
 
-[Stats]
-console_stats_enabled   = true
+; [Stats]
+; console_stats_enabled   = true
```
Comment (nemanja-boric-sociomantic) [inline doc/etc/config.ini]: I think you can just remove this entire section.

Comment (leandro-lucarella-sociomantic) [reply doc/etc/config.ini]: I guess it doesn't hurt to have it explicit since is a file that's supposed to be documentation too.

File: doc/etc/config.ini (hunk 2 above)
Comment (matthias-wende-sociomantic) [reply doc/etc/config.ini]: I'll remove them here as they are still in the file above the section.

File: docker/Dockerfile.dmqnode (hunk 4 above)
Comment (matthias-wende-sociomantic) [reply docker/Dockerfile.dmqnode]: It's not a copy paste. This runs the beaver script for setting the `DMD` env variable. What looks like copied here sets the `PKG` variable for installing the right compiler package.

File: doc/etc/config.ini (hunk 6 above)
Comment (matthias-wende-sociomantic) [reply doc/etc/config.ini]: I'll do although tot entirely related here ;).

File: docker/Dockerfile.dmqnode (hunk 4 above)
Comment (leandro-lucarella-sociomantic) [reply docker/Dockerfile.dmqnode]: But you say it is not (copied)?

File: docker/Dockerfile.dmqnode [hunk 7]
```diff
@@ -1,7 +1,7 @@
 # Duplicate definition of DIST before FROM is needed to be able to use it
 # in docker image name:
 ARG  DIST=xenial
-FROM sociomantictsunami/develdlang:$DIST-v6 as builder
+FROM sociomantictsunami/develdlang:$DIST-v7 as builder
```
Comment (leandro-lucarella-sociomantic) [inline docker/Dockerfile.dmqnode]: I guess this change is OK but it probably deserves a separate commit?

File: docker/Dockerfile.dmqnode [hunk 8]
```diff
@@ -14,13 +14,17 @@ RUN docker/build.sh
 ARG DIST=xenial
 # For now plain ubuntu image is used as base for simplicity. It certainly can be
 # optimized to bare minimum but right now it is not important:
-FROM sociomantictsunami/runtimebase:$DIST-v6
+FROM sociomantictsunami/runtimebase:$DIST-v7
```
Comment (leandro-lucarella-sociomantic) [inline docker/Dockerfile.dmqnode]: Same

File: .travis.yml [hunk 9]
```diff
@@ -64,7 +64,7 @@ jobs:
         - stage: Test
           script:
               - docker build --build-arg DMD=2.071.2.s* --build-arg DIST=xenial
-                -t dmqnode -f docker/Dockerfile.dmqnode .
+                -t dmqnode .
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml]: This is usually [not recommended](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#understand-build-context), as you'll copy the whole project directory to the docker container context. For example the whole `.git` DB (and temporary tests files that you might have). We started putting all in the root dir and then moved to a subdirectory.

File: docker/Dockerfile.dmqnode [hunk 10]
```diff
@@ -14,13 +14,17 @@ RUN docker/build.sh
...
+
+# Set up directories and install the node
 COPY --from=builder /project/build/production/pkg/ /packages/
-COPY docker/install.sh /tmp/
-RUN /tmp/install.sh && rm /tmp/install.sh
-# Update node folder with relevant config to finalize installation
-COPY ./doc/etc/config.docker.ini /srv/dmqnode/dmqnode-0/etc/config.ini
+COPY docker/install.sh /
+RUN /install.sh && rm /install.sh
```
Comment (leandro-lucarella-sociomantic) [inline docker/Dockerfile.dmqnode]: Here is where I think it might be nice to have the important info easy to see, you could do something like:
```dockerfile
RUN /install.sh pkg1 pkg2 pk3 && rm /install.sh
```
But is just a suggestion in case you wanted to gather all the important info in the dockerfile and leave `install.sh` only with the very uninteresting stuff.

File: .travis.yml (hunk 9 above)
Comment (matthias-wende-sociomantic) [reply .travis.yml]: I see what you mean, but what is the difference here. In both cases the build context is the same. Only the location of the Dockerfile changed.
And honestly I don't see how this could be improved, since almost the whole directory structure is needed for building the node. I'm going to add a `.dockerignore` though, to blacklist some files.

File: docker/Dockerfile.dmqnode (hunk 10 above)
Comment (matthias-wende-sociomantic) [reply docker/Dockerfile.dmqnode]: I think it's fine as it is. One has to look into the file in any case to understand what's going on. And `install.sh` really only installs the node and sets up some dirs.

File: .travis.yml (hunk 9 above)
Comment (matthias-wende-sociomantic) [reply .travis.yml]: Also makd seems to needs the `.git ` directory.

Comment (leandro-lucarella-sociomantic) [reply .travis.yml]: You are right, sorry for the noise.

**Comment by matthias-wende-sociomantic** [comment]
@mihails-strasuns do you anything to add?
